keps/sig-storage/20191125-bucket-provisioning.md

```diff
@@ -0,0 +1,428 @@
+    gcs: [12]
+      bucketName:
+      privateKeyName:
+      projectId:
+      serviceAccount:
+  parameters: [13]
+status:
+  message: [14]
+  phase: [15]
+  conditions:
+```
+
+1. `name`: For Greenfield, generated in the pattern of `bucket-<my-bucket>-<my-bucket-namespace>`. 
+1. `labels`: should be added by controller..  Key’s value should be the provisioner name. Characters that do not adhere to [Kubernetes label conventions](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set) will be converted to ‘-’.
+1. `finalizers`: should be added by controller to defer `Bucket` deletion until backend deletion ops succeed.
+1. `provisioner`: The provisioner field defined in the `BucketClass`.  Used by sidecars to filter Buckets.
+1. `releasePolicy`: Prescribes outcome of a Delete events. **Note:** In Brownfield cases, backing storage instance is Retained, and never deleted.
+   - _Delete_:  the bucket and its contents are destroyed.
+   - _Retain_:  the bucket and its data are preserved with only abstracting Kubernetes being destroyed.
+1. `anonymousAccessMode`:  ACL specifying *uncredentialed* access to the Bucket.  This is applicable for cases where the storage instance or objects are intended to be publicly readable and/or writable.  If `BucketClass.anonymousAccessMode` is set, the value is copied here.  Accepted values:
+   - `private`: Default, disallow uncredentialed access to the storage instance.
+   - `ro`: Read only, uncredentialed users are permitted *read* operations for objects within the storage instance.
+   - `rw`: Read/Write, uncredentialed users are permitted *read and write* operations for objects within the storage instance. 
+   - `wo`: Write, uncredentialed users are permitted *write* operations for objects within the storage instance. 
+1. `bucketClassName`: Name of the associated BucketClass.
+1. `permittedNamespaces`: An array of namespaces, identified by a name and uid, from which  BucketRequests are allowed to bind to the Bucket.  Provided to allow admins a layer of cluster-layer access control.  Does **not** reflect or alter the backing storage instances' ACLs or IAM policies.
+   - In Greenfield, the originating BuckerRequest’s namespace must be specified at time of Bucket generation.
+   - In Brownfield, this list is defined by the admin.
+1. `protocol`: The protocol the application will use to access the storage instance.
+   - `protocolSignature`: Specifies the protocol targeted by this Bucket instance.  One of:
+     - `azureBlob`: data required to target a provisioned azure container and/or storage account.
+     - `s3`: data required to target a provisioned S3 bucket and/or user.
+     - `gcs`: data required to target a provisioned GCS bucket and/or service account.
+1. `parameters`: a copy of the BucketClass parameters.
+1. `message`: a human readable description detailing the reason for the current `phase``.
+1. `phase`: is the current state of the Bucket:
```

Comment:
probably we need some _xyzPending_ phases added.

Reply:
Need to have a longer discussion about this. Lets talk about it in tomorrow's meeting.

Reply:
Created
Bound
Released
ErrXYZ